inference/cluster_store.py

```diff
@@ -24,12 +83,75 @@ def average_similarity(self):
         # TODO 
         pass
 
-    # Function to calculate average distance of current clusters 
+    '''
+   :param entity1: X
+   :param entity2: Y
+   :return: euclidian distance between two entity vector (3D) X and Y
+   '''
+    def euclidian_dist(self,entity1, entity2):
+        X = self.original_data[entity1]
+        Y = self.original_data[entity2]
+        return math.sqrt(pow((X[0]-Y[0]),2) + pow((X[1]-Y[1]),2) + pow((X[2]-Y[2]),2))
+
+    '''
+    :param k1: cluster1
+    :param k2: cluster2
+    :return: average distance between two cluster
+    :formula: dist(Ki,Kj) = mean(dist(eil,ejm)) where every eil belongs to Ki not belongs to Kj
+                                                and every ejk belongs to Kj not belongs to Ki
+    '''
+    def avg_distance_between_two_cluster(self,k1,k2):
+        dist = []
+        for k1_entity in k1:
+            for k2_entity in k2:
+                dist.append(self.euclidian_dist(k1_entity,k2_entity))
+        return statistics.mean(dist)
+
+    # Function to calculate average distance of current clusters
+    # :return: average distance of all clusters store in avg_distance
     def average_distance(self):
-        # TODO
-        pass
+        dist = []
+        for i in range(self.total_cluster):
+            for j in range(i+1, self.total_cluster):
+                dist.append(self.avg_distance_between_two_cluster(self.clusters[i],self.clusters[j]))
+        self.avg_distance = statistics.mean(dist)
+
+    # map cluster with entity name and return the clusters
+    def get_clusters_on_entity_map(self):
+        return_ = self.clusters
+        for cluster in range(self.total_cluster):
+            entitys = [self.entity_map[x] for x in self.clusters[cluster]]
+            return_[cluster]= entitys
+        return return_
 
-    # Function to reset clusters back to original data 
+    # Function to reset clusters back to original data
+    # :reset: clustered flag, total_cluster, avg_distance
+    # :delete: clusters data
     def reset(self):
-        # TODO
-        pass
\ No newline at end of file
+        self.clustered = False
+        self.total_cluster = 0
+        self.clusters = None
+        self.avg_distance = 0.0
+
+if __name__=='__main__':
+    data = np.random.uniform(0,500, [200,3])
```

Comment:
After you refactor `euclidian_dist` to use a loop, and test with this example, let's go ahead and remove these lines as well. I can see that the rest of the code is working!